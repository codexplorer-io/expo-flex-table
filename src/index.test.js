import React, { useState } from 'react';
import { shallow } from 'enzyme';
import {
    Menu,
    IconButton,
    Provider as PaperProvider
} from 'react-native-paper';
import noop from 'lodash/noop';
import constant from 'lodash/constant';
import { FlexTable } from './index';

jest.mock('react', () => ({
    ...jest.requireActual('react'),
    useState: jest.fn()
}));

const renderComponent = props => (
    <PaperProvider>
        <FlexTable {...props} />
    </PaperProvider>
);

const defaultProps = {
    columns: [
        {
            key: 'col1',
            value: 'Column 1'
        },
        {
            key: 'col2',
            value: 'Column 2'
        }
    ],
    rows: [
        {
            key: 'row1',
            cells: [
                {
                    key: 'row1_cell1',
                    value: 'Row 1 Cell 1'
                },
                {
                    key: 'row1_cell2',
                    value: 'Row 1 Cell 2'
                }
            ]
        },
        {
            key: 'row2',
            cells: [
                {
                    key: 'row2_cell1',
                    value: 'Row 2 Cell 1'
                },
                {
                    key: 'row2_cell2',
                    value: 'Row 2 Cell 2'
                }
            ]
        },
        {
            key: 'row3',
            cells: [
                {
                    key: 'row3_cell1',
                    value: 'Row 3 Cell 1'
                },
                {
                    key: 'row3_cell2',
                    value: 'Row 3 Cell 2'
                }
            ]
        }
    ]
};

describe('FlexTable', () => {
    const mockUseState = ({
        isActionsMenuVisible = false,
        setIsActionsMenuVisible = noop
    }) => {
        useState.mockReturnValue([
            isActionsMenuVisible,
            setIsActionsMenuVisible
        ]);
    };

    beforeEach(() => {
        mockUseState({});
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    describe('Without actions', () => {
        it('should render as expected', () => {
            // eslint-disable-next-line lodash/prefer-lodash-method
            const wrapper = shallow(renderComponent(defaultProps))
                .find(FlexTable)
                .dive();

            expect(wrapper).toMatchSnapshot();
            // eslint-disable-next-line lodash/prefer-lodash-method
            wrapper.find('ActionsCell').forEach(actionCell => {
                expect(actionCell.dive()).toMatchSnapshot();
            });
        });
    });

    describe('With actions', () => {
        const handleOnPress = jest.fn();
        const props = {
            ...defaultProps,
            rows: [
                ...defaultProps.rows,
                {
                    key: 'row4',
                    cells: [
                        {
                            key: 'row4_cell1',
                            value: 'Row 4 Cell 1',
                            shouldDisplay: false
                        },
                        {
                            key: 'row4_cell2',
                            value: 'Row 4 Cell 2'
                        }
                    ]
                }
            ],
            getRowActions: row => ({
                actions: [
                    {
                        key: 'alert',
                        getIcon: () => row.key === 'row1' ? 'account' : 'account-alert',
                        getTitle: constant('Row 3 Action 1'),
                        onPress: () => handleOnPress(row)
                    },
                    ...(row.key !== 'row3' ? [] : [
                        {
                            key: 'alert',
                            getIcon: constant('account-alert-outline'),
                            getTitle: constant('Row 3 Action 2'),
                            onPress: () => handleOnPress(row)
                        }
                    ])
                ],
                shouldDisplay: row.key !== 'row4'
            })
        };

        it('should render as expected', () => {
            // eslint-disable-next-line lodash/prefer-lodash-method
            const wrapper = shallow(renderComponent(props))
                .find(FlexTable)
                .dive();

            expect(wrapper).toMatchSnapshot();
            // eslint-disable-next-line lodash/prefer-lodash-method
            wrapper.find('ActionsCell').forEach(actionCell => {
                expect(actionCell.dive()).toMatchSnapshot();
            });
        });

        it('should call row single action onPress', () => {
            // eslint-disable-next-line lodash/prefer-lodash-method
            const wrapper = shallow(renderComponent(props))
                .find(FlexTable)
                .dive();
            // eslint-disable-next-line lodash/prefer-lodash-method
            wrapper.find('ActionsCell').forEach((actionCell, index) => {
                // eslint-disable-next-line lodash/prefer-lodash-method
                index < 2 && actionCell.dive().find(IconButton).props().onPress();
            });

            expect(handleOnPress).toHaveBeenCalledTimes(2);
            expect(handleOnPress).toHaveBeenNthCalledWith(1, props.rows[0]);
            expect(handleOnPress).toHaveBeenNthCalledWith(2, props.rows[1]);
        });

        it('should have menu visible when state is true', () => {
            mockUseState({ isActionsMenuVisible: true });
            // eslint-disable-next-line lodash/prefer-lodash-method
            const wrapper = shallow(renderComponent(props))
                .find(FlexTable)
                .dive();

            // eslint-disable-next-line lodash/prefer-lodash-method
            const isVisible = wrapper.find('ActionsCell')
                .at(2)
                .dive()
                .find(Menu)
                .prop('visible');

            expect(isVisible).toBe(true);
        });

        it('should call action to open menu when multiaction is clicked', () => {
            const setIsActionsMenuVisible = jest.fn();
            mockUseState({ setIsActionsMenuVisible });
            // eslint-disable-next-line lodash/prefer-lodash-method
            const wrapper = shallow(renderComponent(props))
                .find(FlexTable)
                .dive();

            shallow(
                // eslint-disable-next-line lodash/prefer-lodash-method
                wrapper.find('ActionsCell')
                    .at(2)
                    .dive()
                    .find(Menu)
                    .prop('anchor')
            ).props().onPress();

            expect(handleOnPress).not.toHaveBeenCalled();
            expect(setIsActionsMenuVisible).toHaveBeenCalledTimes(1);
            expect(setIsActionsMenuVisible).toHaveBeenCalledWith(true);
        });

        it('should call actions onPress on multiaction row', () => {
            const setIsActionsMenuVisible = jest.fn();
            mockUseState({ setIsActionsMenuVisible });
            // eslint-disable-next-line lodash/prefer-lodash-method
            const wrapper = shallow(renderComponent(props))
                .find(FlexTable)
                .dive();

            // eslint-disable-next-line lodash/prefer-lodash-method
            wrapper.find('ActionsCell')
                .at(2)
                .dive()
                .find(Menu.Item)
                .forEach(item => item.props().onPress());

            expect(handleOnPress).toHaveBeenCalledTimes(2);
            expect(handleOnPress).toHaveBeenNthCalledWith(1, props.rows[2]);
            expect(handleOnPress).toHaveBeenNthCalledWith(2, props.rows[2]);
        });
    });
});
