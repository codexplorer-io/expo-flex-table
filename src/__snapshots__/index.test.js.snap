// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlexTable With actions should render as expected 1`] = `
<Styled(View)>
  <Styled(View)
    tableStyle={
      Object {
        "borderColor": "black",
      }
    }
  >
    <Styled(View)
      hasHeader={true}
      isHeaderCell={true}
      isLastCell={false}
      key="col1"
      shouldDisplay={true}
      tableStyle={
        Object {
          "borderColor": "black",
        }
      }
    >
      <Styled(Text)>
        Column 1
      </Styled(Text)>
    </Styled(View)>
    <Styled(View)
      hasHeader={true}
      isHeaderCell={true}
      isLastCell={false}
      key="col2"
      shouldDisplay={true}
      tableStyle={
        Object {
          "borderColor": "black",
        }
      }
    >
      <Styled(Text)>
        Column 2
      </Styled(Text)>
    </Styled(View)>
    <Styled(View)
      hasHeader={true}
      isActionCell={true}
      isHeaderCell={true}
      isLastCell={true}
      shouldDisplay={true}
      tableStyle={
        Object {
          "borderColor": "black",
        }
      }
    >
      <withTheme(IconButton) />
    </Styled(View)>
  </Styled(View)>
  <Styled(View)
    tableStyle={
      Object {
        "borderColor": "black",
      }
    }
  >
    <Styled(View)
      hasHeader={true}
      isFirstRow={true}
      isLastCell={false}
      key="row1_cell1"
      shouldDisplay={true}
      shouldDisplayNextRow={true}
      tableStyle={
        Object {
          "borderColor": "black",
        }
      }
    >
      <Styled(Text)
        hasTopMargin={false}
      >
        Row 1 Cell 1
      </Styled(Text)>
    </Styled(View)>
    <Styled(View)
      hasHeader={true}
      isFirstRow={true}
      isLastCell={false}
      key="row1_cell2"
      shouldDisplay={true}
      shouldDisplayNextRow={true}
      tableStyle={
        Object {
          "borderColor": "black",
        }
      }
    >
      <Styled(Text)
        hasTopMargin={false}
      >
        Row 1 Cell 2
      </Styled(Text)>
    </Styled(View)>
    <ActionsCell
      getRowActions={[Function]}
      isFirstRow={true}
      row={
        Object {
          "cells": Array [
            Object {
              "key": "row1_cell1",
              "value": "Row 1 Cell 1",
            },
            Object {
              "key": "row1_cell2",
              "value": "Row 1 Cell 2",
            },
          ],
          "key": "row1",
        }
      }
      tableStyle={
        Object {
          "borderColor": "black",
        }
      }
    />
  </Styled(View)>
  <Styled(View)
    tableStyle={
      Object {
        "borderColor": "black",
      }
    }
  >
    <Styled(View)
      hasHeader={true}
      isFirstRow={false}
      isLastCell={false}
      key="row2_cell1"
      shouldDisplay={true}
      shouldDisplayNextRow={true}
      tableStyle={
        Object {
          "borderColor": "black",
        }
      }
    >
      <Styled(Text)
        hasTopMargin={false}
      >
        Row 2 Cell 1
      </Styled(Text)>
    </Styled(View)>
    <Styled(View)
      hasHeader={true}
      isFirstRow={false}
      isLastCell={false}
      key="row2_cell2"
      shouldDisplay={true}
      shouldDisplayNextRow={true}
      tableStyle={
        Object {
          "borderColor": "black",
        }
      }
    >
      <Styled(Text)
        hasTopMargin={false}
      >
        Row 2 Cell 2
      </Styled(Text)>
    </Styled(View)>
    <ActionsCell
      getRowActions={[Function]}
      isFirstRow={false}
      row={
        Object {
          "cells": Array [
            Object {
              "key": "row2_cell1",
              "value": "Row 2 Cell 1",
            },
            Object {
              "key": "row2_cell2",
              "value": "Row 2 Cell 2",
            },
          ],
          "key": "row2",
        }
      }
      tableStyle={
        Object {
          "borderColor": "black",
        }
      }
    />
  </Styled(View)>
  <Styled(View)
    tableStyle={
      Object {
        "borderColor": "black",
      }
    }
  >
    <Styled(View)
      hasHeader={true}
      isFirstRow={false}
      isLastCell={false}
      key="row3_cell1"
      shouldDisplay={true}
      shouldDisplayNextRow={false}
      tableStyle={
        Object {
          "borderColor": "black",
        }
      }
    >
      <Styled(Text)
        hasTopMargin={false}
      >
        Row 3 Cell 1
      </Styled(Text)>
    </Styled(View)>
    <Styled(View)
      hasHeader={true}
      isFirstRow={false}
      isLastCell={false}
      key="row3_cell2"
      shouldDisplay={true}
      shouldDisplayNextRow={true}
      tableStyle={
        Object {
          "borderColor": "black",
        }
      }
    >
      <Styled(Text)
        hasTopMargin={false}
      >
        Row 3 Cell 2
      </Styled(Text)>
    </Styled(View)>
    <ActionsCell
      getRowActions={[Function]}
      isFirstRow={false}
      row={
        Object {
          "cells": Array [
            Object {
              "key": "row3_cell1",
              "value": "Row 3 Cell 1",
            },
            Object {
              "key": "row3_cell2",
              "value": "Row 3 Cell 2",
            },
          ],
          "key": "row3",
        }
      }
      tableStyle={
        Object {
          "borderColor": "black",
        }
      }
    />
  </Styled(View)>
  <Styled(View)
    tableStyle={
      Object {
        "borderColor": "black",
      }
    }
  >
    <Styled(View)
      hasHeader={true}
      isFirstRow={false}
      isLastCell={false}
      key="row4_cell1"
      shouldDisplay={false}
      shouldDisplayNextRow={false}
      tableStyle={
        Object {
          "borderColor": "black",
        }
      }
    />
    <Styled(View)
      hasHeader={true}
      isFirstRow={false}
      isLastCell={false}
      key="row4_cell2"
      shouldDisplay={true}
      shouldDisplayNextRow={false}
      tableStyle={
        Object {
          "borderColor": "black",
        }
      }
    >
      <Styled(Text)
        hasTopMargin={false}
      >
        Row 4 Cell 2
      </Styled(Text)>
    </Styled(View)>
    <ActionsCell
      getRowActions={[Function]}
      isFirstRow={false}
      row={
        Object {
          "cells": Array [
            Object {
              "key": "row4_cell1",
              "shouldDisplay": false,
              "value": "Row 4 Cell 1",
            },
            Object {
              "key": "row4_cell2",
              "value": "Row 4 Cell 2",
            },
          ],
          "key": "row4",
        }
      }
      tableStyle={
        Object {
          "borderColor": "black",
        }
      }
    />
  </Styled(View)>
</Styled(View)>
`;

exports[`FlexTable With actions should render as expected 2`] = `
<Styled(View)
  isActionCell={true}
  isFirstRow={true}
  isLastCell={true}
  shouldDisplay={true}
  tableStyle={
    Object {
      "borderColor": "black",
    }
  }
>
  <withTheme(IconButton)
    color="#6200ee"
    icon="account"
    onPress={[Function]}
  />
</Styled(View)>
`;

exports[`FlexTable With actions should render as expected 3`] = `
<Styled(View)
  isActionCell={true}
  isFirstRow={false}
  isLastCell={true}
  shouldDisplay={true}
  tableStyle={
    Object {
      "borderColor": "black",
    }
  }
>
  <withTheme(IconButton)
    color="#6200ee"
    icon="account-alert"
    onPress={[Function]}
  />
</Styled(View)>
`;

exports[`FlexTable With actions should render as expected 4`] = `
<Styled(View)
  isActionCell={true}
  isFirstRow={false}
  isLastCell={true}
  shouldDisplay={true}
  tableStyle={
    Object {
      "borderColor": "black",
    }
  }
>
  <withTheme(Menu)
    anchor={
      <withTheme(IconButton)
        color="#6200ee"
        icon="menu-down-outline"
        onPress={[Function]}
      />
    }
    onDismiss={[Function]}
    visible={false}
  >
    <withTheme(Menu.Item)
      key="alert"
      onPress={[Function]}
      title="Row 3 Action 1"
    />
    <withTheme(Menu.Item)
      key="alert"
      onPress={[Function]}
      title="Row 3 Action 2"
    />
  </withTheme(Menu)>
</Styled(View)>
`;

exports[`FlexTable With actions should render as expected 5`] = `
<Styled(View)
  isActionCell={true}
  isFirstRow={false}
  isLastCell={true}
  shouldDisplay={false}
  tableStyle={
    Object {
      "borderColor": "black",
    }
  }
>
  <withTheme(IconButton) />
</Styled(View)>
`;

exports[`FlexTable Without actions should render as expected 1`] = `
<Styled(View)>
  <Styled(View)
    tableStyle={
      Object {
        "borderColor": "black",
      }
    }
  >
    <Styled(View)
      hasHeader={true}
      isHeaderCell={true}
      isLastCell={false}
      key="col1"
      shouldDisplay={true}
      tableStyle={
        Object {
          "borderColor": "black",
        }
      }
    >
      <Styled(Text)>
        Column 1
      </Styled(Text)>
    </Styled(View)>
    <Styled(View)
      hasHeader={true}
      isHeaderCell={true}
      isLastCell={true}
      key="col2"
      shouldDisplay={true}
      tableStyle={
        Object {
          "borderColor": "black",
        }
      }
    >
      <Styled(Text)>
        Column 2
      </Styled(Text)>
    </Styled(View)>
  </Styled(View)>
  <Styled(View)
    tableStyle={
      Object {
        "borderColor": "black",
      }
    }
  >
    <Styled(View)
      hasHeader={true}
      isFirstRow={true}
      isLastCell={false}
      key="row1_cell1"
      shouldDisplay={true}
      shouldDisplayNextRow={true}
      tableStyle={
        Object {
          "borderColor": "black",
        }
      }
    >
      <Styled(Text)
        hasTopMargin={false}
      >
        Row 1 Cell 1
      </Styled(Text)>
    </Styled(View)>
    <Styled(View)
      hasHeader={true}
      isFirstRow={true}
      isLastCell={true}
      key="row1_cell2"
      shouldDisplay={true}
      shouldDisplayNextRow={true}
      tableStyle={
        Object {
          "borderColor": "black",
        }
      }
    >
      <Styled(Text)
        hasTopMargin={false}
      >
        Row 1 Cell 2
      </Styled(Text)>
    </Styled(View)>
    <ActionsCell
      isFirstRow={true}
      row={
        Object {
          "cells": Array [
            Object {
              "key": "row1_cell1",
              "value": "Row 1 Cell 1",
            },
            Object {
              "key": "row1_cell2",
              "value": "Row 1 Cell 2",
            },
          ],
          "key": "row1",
        }
      }
      tableStyle={
        Object {
          "borderColor": "black",
        }
      }
    />
  </Styled(View)>
  <Styled(View)
    tableStyle={
      Object {
        "borderColor": "black",
      }
    }
  >
    <Styled(View)
      hasHeader={true}
      isFirstRow={false}
      isLastCell={false}
      key="row2_cell1"
      shouldDisplay={true}
      shouldDisplayNextRow={true}
      tableStyle={
        Object {
          "borderColor": "black",
        }
      }
    >
      <Styled(Text)
        hasTopMargin={false}
      >
        Row 2 Cell 1
      </Styled(Text)>
    </Styled(View)>
    <Styled(View)
      hasHeader={true}
      isFirstRow={false}
      isLastCell={true}
      key="row2_cell2"
      shouldDisplay={true}
      shouldDisplayNextRow={true}
      tableStyle={
        Object {
          "borderColor": "black",
        }
      }
    >
      <Styled(Text)
        hasTopMargin={false}
      >
        Row 2 Cell 2
      </Styled(Text)>
    </Styled(View)>
    <ActionsCell
      isFirstRow={false}
      row={
        Object {
          "cells": Array [
            Object {
              "key": "row2_cell1",
              "value": "Row 2 Cell 1",
            },
            Object {
              "key": "row2_cell2",
              "value": "Row 2 Cell 2",
            },
          ],
          "key": "row2",
        }
      }
      tableStyle={
        Object {
          "borderColor": "black",
        }
      }
    />
  </Styled(View)>
  <Styled(View)
    tableStyle={
      Object {
        "borderColor": "black",
      }
    }
  >
    <Styled(View)
      hasHeader={true}
      isFirstRow={false}
      isLastCell={false}
      key="row3_cell1"
      shouldDisplay={true}
      shouldDisplayNextRow={false}
      tableStyle={
        Object {
          "borderColor": "black",
        }
      }
    >
      <Styled(Text)
        hasTopMargin={false}
      >
        Row 3 Cell 1
      </Styled(Text)>
    </Styled(View)>
    <Styled(View)
      hasHeader={true}
      isFirstRow={false}
      isLastCell={true}
      key="row3_cell2"
      shouldDisplay={true}
      shouldDisplayNextRow={false}
      tableStyle={
        Object {
          "borderColor": "black",
        }
      }
    >
      <Styled(Text)
        hasTopMargin={false}
      >
        Row 3 Cell 2
      </Styled(Text)>
    </Styled(View)>
    <ActionsCell
      isFirstRow={false}
      row={
        Object {
          "cells": Array [
            Object {
              "key": "row3_cell1",
              "value": "Row 3 Cell 1",
            },
            Object {
              "key": "row3_cell2",
              "value": "Row 3 Cell 2",
            },
          ],
          "key": "row3",
        }
      }
      tableStyle={
        Object {
          "borderColor": "black",
        }
      }
    />
  </Styled(View)>
</Styled(View)>
`;

exports[`FlexTable Without actions should render as expected 2`] = `""`;

exports[`FlexTable Without actions should render as expected 3`] = `""`;

exports[`FlexTable Without actions should render as expected 4`] = `""`;
